### ISSUE TITLE
Requests

### ISSUE THREAD

**Issue by Anbranin** [umts/chronos#5]
Time off requests
attributes:
requested_by (user)
start-date-time
end-date-time
type - PER, VAC, SCK, FSK (Check with Metin there might be more)
approved - boolean
approved_by - supervisor

Need the ability to view these time off requests in a calendar along with the schedule

**Comment by nawilliams** [comment]
You can see what the current one looks like on the old hub under calendar. https://parking.umass.edu/hotspot.php/calendarcontroller

**Comment by nawilliams** [comment]
I'd recommend using either [simple calendar](https://github.com/excid3/simple_calendar) or [full calendar-rails](https://github.com/bokmann/fullcalendar-rails) for the display part.

**Comment by nawilliams** [comment]
General view should show the schedules of everyone. If the current user is a supervisor then also show the requests that are from users that are supervisees. We don't have a permission system yet, so don't worry too much about this.

**Comment by LiamBrandt** [comment]
I will continue working on requests with these points in mind:
* A time off request object should not be able to span multiple days. The form for time off request creation will allow a date range, time range, days of the week picker. A time off request will be created individually for each day, and each request will have to be approved by a supervisor on a day to day basis.
The fields will look similar to this old form:
![screen shot 2018-04-26 at 12 51 58 pm](https://user-images.githubusercontent.com/7032019/39319959-ac6d1b6a-4950-11e8-8527-d913a9c2939e.png)

* Supervisors should not be able to approve or reject their own time off requests, only their supervisors can.
* There are three statuses for time off requests, approved, pending or rejected. On creation they are pending, only shown to supervisors. Once approved, they can be switched back to pending, or rejected (not shown at all).
* You should not be able to update a time off request after creation, only ask that a supervisor delete it (reject it) so you can re add it with changes.
* Supervisors will have a separate tab to approve requests of their nested subordinates.
* There can be chains of supervisors, and one user should be able to have multiple supervisors.
* Admins should be able to act as supervisors for any user.
* Nothing should be deleted, just deactivated (in this case rejected).

**Comment by Anbranin** [comment]
I disagree with one point:
> You should not be able to update a time off request after creation, only ask that a supervisor delete it (reject it) so you can re add it with changes.

Would it be OK if we allowed editing up until the point at which it was approved? If I made a typo and accidentally add one more day, I see no problem with being able to edit the request if it's in the `pending` status.
The controller action could just check to see if the status is pending and if it is, allow editing.

**Comment by Anbranin** [comment]
Also, I hope the new form looks a _little_ different at least from the old one. Having to toggle between time range and shifts like that doesn't look really clean.

**Comment by LiamBrandt** [comment]
@Anbranin Yes, the form will look different. For editing while pending until approval you will have to ask @umayavuz. I think part of the problem is that people will be requesting time off on the touchscreen interface, and it might be harder to navigate a calendar, find your time off, then edit it again.

The current way to request time off is here:
https://parking.umass.edu/hotspot.php/timemachine
Be warned that you click buttons by just hovering the mouse, I assume so that you cannot click and navigate out of the browser to another page. It doesnt seem like you can edit requests after creation.

**Comment by Anbranin** [comment]
Yeah I didn't consider the whole touchscreen thing. I guess we should just go with very basic functionality for now and if it's requested we can add editing.

**Comment by Anbranin** [comment]
Actually I take it back, it might be hard to do it on a touchscreen, but you _can_ go to a web browser and do it. I'll ask Metin.
